File: Orimath/ViewModels/SettingViewModel.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Reflection;
using System.Threading.Tasks;
using Orimath.Controls;

namespace Orimath.ViewModels
{
    public class SettingViewModel
    {
        private readonly object _obj;
        private bool _loaded;
        private readonly ResettableObservableCollection<SettingItemViewModel> _items = new();

        public ResettableObservableCollection<SettingItemViewModel> Items
        {
            get
            {
                if (!_loaded) LoadItems();
                return _items;
            }
        }

        public SettingViewModel(object obj)
        {
            _obj = obj;
        }

        private async void LoadItems()
        {
            _loaded = true;

            var values = await Task.Run(() =>
            {
                var result = new List<SettingItemViewModel>();

                foreach (var prop in _obj.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance))
                {
                    if (prop.GetGetMethod() is null || prop.GetSetMethod() is null)
                        continue;

                    if (prop.GetCustomAttribute<EditableAttribute>() is { } editable &&
                        !editable.AllowEdit)
                        continue;

                    var propertyType = prop.PropertyType;

                    if (propertyType == typeof(double))
                        result.Add(new DoubleSettingItemViewModel(prop, _obj));
                    else if (propertyType == typeof(int))
                        result.Add(new Int32SettingItemViewModel(prop, _obj));
                    else if (propertyType == typeof(bool))
                        result.Add(new BooleanSettingItemViewModel(prop, _obj));
                    else if (propertyType == typeof(string))
                        result.Add(new StringSettingItemViewModel(prop, _obj));
                    else if (propertyType.IsEnum)
                        result.Add(new EnumSettingItemViewModel(prop, _obj));
                }

                return result;
            });

            _items.Reset(values);
        }
    }
}

File: Orimath/ViewModels/SettingViewModel.fs
namespace Orimath.ViewModels
open System.ComponentModel.DataAnnotations
open System.Reflection
open Orimath
open Orimath.Controls
open Orimath.Plugins
open Orimath.Internal

type SettingViewModel(object: obj, dispatcher: IDispatcher) =
    let mutable loaded = false
    let items = ResettableObservableCollection<SettingItemViewModel>()

    member this.Items =
        if not loaded then this.LoadItems()
        items

    member private _.LoadItems() =
        loaded <- true
        dispatcher.Background {
            let result = [|
                for prop in object.GetType().GetProperties(BindingFlags.Public ||| BindingFlags.Instance) do
                if isNotNull (prop.GetGetMethod()) && isNotNull (prop.GetSetMethod()) &&
                    (prop.GetCustomAttribute<EditableAttribute>()
                    |> Null.mapv(fun att -> att.AllowEdit)
                    |> Option.defaultValue true) then
                    let propertyType = prop.PropertyType
                    if propertyType = typeof<float> then
                        yield DoubleSettingItemViewModel(prop, object) :> SettingItemViewModel
                    elif propertyType = typeof<int> then
                        yield upcast Int32SettingItemViewModel(prop, object)
                    elif propertyType = typeof<bool> then
                        yield upcast BooleanSettingItemViewModel(prop, object)
                    elif propertyType = typeof<string> then
                        yield upcast StringSettingItemViewModel(prop, object)
                    elif propertyType.IsEnum then
                        yield upcast EnumSettingItemViewModel(prop, object) |]
            do! dispatcher.UI
            items.Reset(result)
        }
